crates/optimism/trie/src/db/store.rs

```diff
@@ -327,12 +367,86 @@ impl OpProofsStore for MdbxProofsStorage {
         unimplemented!()
     }
 
+    /// Prune all historical trie data prior to `new_earliest_block_number` using
+    /// the [`BlockChangeSet`] index.
     async fn prune_earliest_state(
         &self,
-        _new_earliest_block_number: u64,
+        new_earliest_block_number: u64,
         _diff: BlockStateDiff,
     ) -> OpProofsStorageResult<()> {
-        unimplemented!()
+        let earliest_block = self.get_earliest_block_number().await?;
+        let start_block = if let Some((block, _)) = earliest_block {
+            if block >= new_earliest_block_number {
+                return Ok(()) // Nothing to prune
+            }
+            block
+        } else {
+            return Ok(()) // Nothing to prune
+        };
+
+        self.env.update(|tx| {
+            // Collect keys to prune first to avoid borrow checker issues with cursors.
+            let keys_to_prune: Vec<(u64, ChangeSet)> = {
+                let mut change_set_cursor = tx.new_cursor::<BlockChangeSet>()?;
+                change_set_cursor
+                    .walk_range(start_block..new_earliest_block_number)?
+                    .collect::<Result<Vec<_>, _>>()?
+            };
+
+            if !keys_to_prune.is_empty() {
+                let mut account_trie_cursor = tx.new_cursor::<AccountTrieHistory>()?;
+                let mut storage_trie_cursor = tx.new_cursor::<StorageTrieHistory>()?;
+                let mut hashed_account_cursor = tx.new_cursor::<HashedAccountHistory>()?;
+                let mut hashed_storage_cursor = tx.new_cursor::<HashedStorageHistory>()?;
+                let mut change_set_cursor = tx.new_cursor::<BlockChangeSet>()?;
+
+                // Process already sorted entries directly
+                for (block_number, change_set) in &keys_to_prune {
+                    // Process account trie entries
+                    for key in &change_set.account_trie_keys {
+                        if account_trie_cursor
+                            .seek_by_key_subkey(key.clone(), *block_number)?
+                            .is_some()
+                        {
+                            account_trie_cursor.delete_current()?;
+                        }
+                    }
+
+                    // Process storage trie entries
+                    for key in &change_set.storage_trie_keys {
+                        if storage_trie_cursor
+                            .seek_by_key_subkey(key.clone(), *block_number)?
+                            .is_some()
+                        {
+                            storage_trie_cursor.delete_current()?;
+                        }
+                    }
+
+                    // Process hashed account entries
+                    for key in &change_set.hashed_account_keys {
+                        if hashed_account_cursor.seek_by_key_subkey(*key, *block_number)?.is_some()
```

Comment:
Since `seek_by_key_subkey` returns entry greater than or equal to the provided key/subkey pair, I think we should cross check the key and block number here with the data returned.